### DISCUSSION TITLE
About operations and tmatrix class

### DISCUSSION THREAD

**Discussion by jgenee** [thelfer/tfel#165] [Q&A] [answered]
Hello everyone,

I am a relatively new user of Mfront, so the question might be relatively easy to answer. Though, I was wondering if operations between nonsquare matrix (NxM with N and M different) are handled within Mfront (multiplication, etc). 

For example, I defined two matrices A and B as :

@LocalVariable tfel::math::tmatrix<9,18,real> A; 
@LocalVariable tfel::math::tmatrix<18,9,real> B;

I need to calculate something like B = invert(transpose(A)*A) * transpose(A).
I tried this the hard way but it does not seem to work, as there is no match for " operator * ". I guess the problem will be the same for the invert function..

Am I missing something ?

Thanks in advance !

**Comment by thelfer** [comment, marked as answer]
@jgenee, Thanks for your interest in `TFEL/MFront`. You did not miss anything. It just seems that I never needed to implement the matrix product. In pratice, I have implemented the product of tensors. Maybe you could give me more insight on what your are trying to do just to figure out why you need it. The standard implementation is easy to add.

**Reply by thelfer** [reply]
You can add the code below in your `MFront` file (usually somewhere near the top). I did not cheked it, so you have to :)

Then you can use it as follows:

```cxx
const auto B = tfel::math::extensions::my_complex_operation(A);
```

Best,
Thomas

```cxx
@Includes {

#ifndef JGENEE_TFEL_MATH_EXTENSIONS
#define JGENEE_TFEL_MATH_EXTENSIONS 1

namespace tfel::math::extensions {

template <unsigned short N, unsigned short M, typename real>
auto tranpose_multiply(const tmatrix<N, M, real> &A) {
  auto r = tmatrix<M, M, result_type<real, real, OpMult>>{};
  for (index_type<tmatrix<N, M, real>> i = 0; i != N; ++i) {
    for (index_type<tmatrix<N, M, real>> j = 0; j != N; ++j) {
      r(i, j) = real{0} * real{0};
      for (index_type<tmatrix<N, M, real>> k = 0; k != M; ++k) {
        r(i, j) += A(k, i) * A(k, j);
      }
    }
  }
  return r;
}

template <unsigned short N, typename real>
std::pair<bool, tmatrix<N, N, real>> invert(tmatrix<N, N, real> &m) {
  //  static_assert(std::is_same_v<real, invert_type<real>>);
  auto im = tmatrix<N, N, real>::Id();
  auto success = TinyMatrixSolve<N, real, false>::exe(m, im);
  return {success, im};
}

template <unsigned short N, unsigned short M, typename real>
std::pair<bool, tmatrix<M, N, real>>
my_complex_operation(const tmatrix<N, M, real> &A) {
  auto r = tmatrix<M, N, result_type<real, real, OpMult>>{};
  auto tmp2 = tranpose_multiply(A);
  const auto tmp = invert(tmp2); // M x M
  if (!tmp.first) {
    return {false, r};
  }
  const auto &im = tmp.second;
  for (index_type<tmatrix<M, N, real>> i = 0; i != M; ++i) {
    for (index_type<tmatrix<M, N, real>> j = 0; j != N; ++j) {
      r(i, j) = real{0} * real{0};
      for (index_type<tmatrix<M, N, real>> k = 0; k != M; ++k) {
        r(i, j) += im(i, k) * A(j, k);
      }
    }
  }
  return {true, r};
}

} // namespace tfel::math::extensions

#endif JGENEE_TFEL_MATH_EXTENSIONS 1

}
```

**Reply by jgenee** [reply]
Thank you so much for your help ! 
I will try and check that ;)

Is there a more esthetic way to add this code ? Maybe save it as a hxx/cxx/ixx file (I am not yet familiar with the extensions) and call it in the main mfront file ?

**Reply by thelfer** [reply]
@jgenee I usually add my extensions in a header file such as `TFELMathExtensions.hxx` and include it as follows:

```cxx
@Includes{
#include "TFELMathExtensions.hxx"
}
```

and then specify to `mfront` where to find the header file:

```bash
$ mfront -I $(pwd) ---obuild ....
```

**Reply by jgenee** [reply]
Great ! Thanks again ;)

**Reply by breou** [reply]
Hello,

I am Breno Nogueira, PhD student at ENS Paris-Saclay and working also with CEA Saclay. I have already worked with Mfront for the microplane damage (which indeed worked quiet nice as we had to deal only with tensors).

Nice to see this post because i am currently facing the same problem in another subject.

I am developing E-FEM method in Mfront (as the formulation is very similar to a local behaviour law at Gauss point level). In this case, i need to introduce some matrices and even solve a linear system (2x2 in 2D and 3x3 in 3D).

I construct (for the 2D case): 

//Construction matrix G=Bc, N and D
  auto G = tmatrix<3, 2, real>{};      // initializing G matrix
  auto N = tmatrix<3, 2, real>{};      // initializing N matrix
  auto Dmatrix = tmatrix<3, 3, stress>{}; // i actually need elasticity tensor in its matrix form for this computation
  auto K = tmatrix<2, 2, real>{};      // initializing G matrix
  
 and then i would need to do :
 
 Z = alpha*K - transpose(N)*(Dmatrix*G); // with alpha real 
 
and solve for U : 

ZU = t; // with U and T two "tvectors" with dimension (2,1).

I suppose that i will also need to include these matrix computations based on your code here above. However, is it possible to solve this system ?

Thanks in advance !

Breno

**Reply by thelfer** [reply]
Hi @breou,

> I have already worked with Mfront for the microplane damage (which indeed worked quiet nice as we had to deal only with tensors).

Seems like you're ready for a talk at the MFront User day !

> I suppose that i will also need to include these matrix computations based on your code here above.

Indeed. You may also ask for appropriate extensions in the `TFEL/Math` library by opening dedicated issues (@jgenee do not hesitate).

> However, is it possible to solve this system ?

Sorry, but I don't see the issue here if `Z` is invertible...
Just use `TinyMatrixSolve<N, real, false>::exe(m, t)` where `t` is a `tvector` containing the values of `U` on input.

**Reply by breou** [reply]
Thanks for the answer,

Sure for the Mfront talk for microplane if it is possible. We hope to do good things in the next years coupling damage behaviour from Mfront and nonlocal gradient solution in PSD (developed in CEA saclay).

And yes Z is invertible, i just didn't know that TinyMatrixSolve exists. Perfect !

I will try to do it. Thx again !

**Reply by thelfer** [reply]
@breou What is PSD ? You can contact me directly on my email for details !

**Reply by SCrenSC** [reply]
Hi Thomas,
I have the same need for tmatrix math operation. I'm glad to find the answer here. However, I had some errors when I tried to compile the file with your .hxx.

1. If I define the B matrix as
@LocalVariable tfel::math::tmatrix<18,9,real> B;
B= tfel::math::extensions::my_complex_operation(A); 
then I will get the following error message:
myfile.mfront:121:14: error: no match for ‘operator=’ (operand types are ‘tfel::math::tmatrix<18, 9, double>’ and ‘std::pair<bool, tfel::math::tmatrix<18, 9, double> >’)

2. If I add  const auto as you suggested in the post
@LocalVariable tfel::math::tmatrix<18,9,real> B;
const auto B= tfel::math::extensions::my_complex_operation(A); 
I will get:
myfile.mfront: In member function ‘bool tfel::material::mymainfile<tfel::material::ModellingHypothesis::TRIDIMENSIONAL, NumericType, false>::initialize()’:
myfile.mfront:121:12: error: expected unqualified-id before ‘this’
compilation terminated due to -Wfatal-errors.


3. only warning message without using @LocalVariable tfel::math::tmatrix<18,9,real> B;
and with const auto B=tfel::math::extensions::my_complex_operation(A);  
.../TFELMathExtensions.hxx:53:8: warning: extra tokens at end of #endif directive [-Wendif-labels]
   53 | #endif SR_TFEL_MATH_EXTENSIONS 1
      |        ^~~~~~~~~~~~~~~~~~~~~~~

Btw, I am using version 4.1.0.
Did I miss anything? I wanted to put the result in tmatrix. Thank you very much for your help.
Sicong

**Reply by thelfer** [reply]
Hi Sicong,

# Question 1

The error message states that:

1. `tfel::math::extensions::my_complex_operation` returns a pair `std::pair<bool, tfel::math::tmatrix<18, 9, double> >`
2. You can't assign this pair to a `tfel::math::tmatrix<18, 9, double>`.

I don't remember any context about `tfel::math::extensions::my_complex_operation` but I guess that the bool part of the pair states of the function succeeded or not.

In your case, you shall (probably, again it's a guess):

```c++
auto success = bool{};
tie(success, B) = tfel::math::extensions::my_complex_operation(...);
if(!success){
  return false; // not sure
}
```

# Question 2

```c++
@localvariable tfel::math::tmatrix<18,9,real> B;
....
const auto B= ...;
```

This can't work because `MFront` adds `this->` to all the variables it manages to alleviate the syntax (adding `this` in front of the variables is requires in some circumstances related to inheritance in `template` context). So your code translates in:

```c++
const auto this->B= ...;
```

which is invalid.

All in all, the error message is awful, but it basically mean that you can't defined a variable with the same name as an `MFront` variable

# Question 3

This is a simple warning, you shall write:

```c++
 #endif /* SR_TFEL_MATH_EXTENSIONS */
 ```

**Reply by SCrenSC** [reply]
Thank you! The following lines are working.
auto success = bool{};
tie(success, B) = tfel::math::extensions::my_complex_operation(...);
if(!success){
  return false; // not sure
}

**Reply by thelfer** [reply]
Great !

**Reply by thelfer** [reply]
Let me know if you have other troubles

**Reply by SCrenSC** [reply]
Some new troubles indeed. 
The compilation was successful, however when I tried the behaviours in Amitex, I got:
The behaviour 'mymatbehavior' has thrown a generic standard exception : initialization of behaviour 'mymatbehavior' failed. 
It seems it is not recognized in Amitex.

**Reply by thelfer** [reply]
You may want try:

```c++
auto success = bool{};
tie(success, B) = tfel::math::extensions::my_complex_operation(...);
if(!success){
  std::cout << "my_complex_operation failed\n";  
  return false; // not sure
}
std::cout << "my_complex_operation was successful\n";
```

**Comment by jgenee** [comment]
Thanks for your answer.

I am implementing a non local behavior law using MFront and Amitex. To that end, I defined a 9x18 matrix (A in the previous message) that contain information about dislocations types. Then, the matrix is used in the calculation of the 18x18 matrix B that relates two state variables in the behavior law (one being an non-local value calculated outside of the behavior law by Amitex). I defined the calculation of B in a second mfront file that I imported into the main one containing the law.

Basically, A is initialized as a 9x18 matrix. And I need to compute B = invert( transpose(A) * A ) * transpose(A).
As you told me, operation * is not defined for matrices, same for invert and transpose functions (not sure about invert though, as it is here used on a square matrix).

Hope I am clear enough !
